API: allow scalar setting/getting via float indexer on integer indexes

jreback opened the pull request:
closes #12333

jreback committed:
API: allow scalar setting/getting via float indexer on integer indexes

closes #12333

jreback commented:
not a lot of code changes, mostly making the tests robust & w/o repeating things.

haven't updated docs yet

jorisvandenbossche commented on pandas/tests/test_indexing.py:
```diff
@@ -5147,16 +5147,13 @@ def check_index(index, error):
-        self.assertEqual(result4, result6)
+        for result in [s[5.0], s[5],
+                       s.loc[5.0], s.loc[5],
+                       s.ix[5.0], s[5]]:
```
the last one here should probably be `s.ix[5]`

jreback replied:
fixed

jorisvandenbossche commented on pandas/tests/test_indexing.py:
```diff
@@ -5619,19 +5650,12 @@ def f():
+                    for idxr in [lambda x: x.loc,
+                                 lambda x: x.iloc,
+                                 lambda x: x]:
+                        self.assertRaises(TypeError, lambda: idxr(s)[l])
```
I am probably missing something (the diff in the tests is a bit hard to follow), but shouldn't there be a difference between loc and iloc here? As loc is doing label-based slicing, so float is allowed?

jreback replied:
these are about slicing:

In [1]: s = Series(range(3))

In [2]: s[1.0:2]
TypeError: cannot do slice start value indexing on <class 'pandas.indexes.range.RangeIndex'> with these indexers [1.0] of <type 'float'>

In [3]: s.loc[1.0:2]
TypeError: cannot do slice indexing on <class 'pandas.indexes.range.RangeIndex'> with these indexers [1.0] of <type 'float'>

In [4]: s.ix[1.0:2]
Out[4]: 
1    1
2    2
dtype: int64
```

jorisvandenbossche replied:
Yes, but why would slicing be an exception?

```
In [107]: s.loc[1.0]
Out[107]: 1

In [108]: s.loc[2.0]
Out[108]: 2

In [109]: s.loc[1.0:2.0]
TypeError: cannot do slice indexing on <class 'pandas.indexes.numeric.Int64Index
'> with these indexers [1.0] of <type 'float'>
```

In ``s[1.0:2]``, the `1.0` is an integer location, so that should indeed raise, but in `s.loc[1.0:2]`, the `1.0` is a label.

jreback replied:
https://github.com/pydata/pandas/issues/12333#issuecomment-185246550

this is the same as your comment issue. We don't allow float indexers in slicing at ALL, except for ``Float64Index``. This is possible to relax, but introduces another layer of complexity here. (this doesn't apply for ``.ix``, just ``[]`` and ``.loc``)

jorisvandenbossche replied:
We perfectly allow floats in slices?

```
In [1]: s = Series(range(3))

In [2]: s.loc[1.0:2]
Out[2]:
1    1
2    2
dtype: int64

In [3]: s.ix[1.0:2]
Out[3]:
1    1
2    2
dtype: int64

In [4]: pd.__version__
Out[4]: u'0.17.1'
```

jreback commented on pandas/indexes/base.py:
```diff
@@ -1889,7 +1889,20 @@ def get_loc(self, key, method=None, tolerance=None):
+                if is_float(key) and not self.is_floating():
+                    try:
+                        ckey = int(key)
+                        if ckey == key:
```
This should actully be done inside the indexing engine......will make an issure for it

jreback commented:
ok, this is updated

- reorged indexing tests (mainly moved out float indexers from the monolit ``test_indexing.py`` and dropped some duplicate tests)
- bulit comprehensive float scalar/slice testing, separated out by index
- removed some net code! yeh
- cleaned up the some duplicated / awkward indexing code (*some*, but it does flow a bit better now). main issue was some duplicated code w.r.t. some of the slice indexing and checking. hopefully this will allow future changes to be in single places rather than all over the place.

net-net-net. I don't think actually much has changed from 0.17.1. (with the exception that I note when setting with ``.ix`` in the docs, IOW, using a float indexer on an object index now sets the label rather than positional sets).

 We have dropped the ``FutureWarning`` entirely on float indexers, but now only raise for ``.iloc`` (previously we were also raising on non-float indexes for ``[]``, ``.ix``, ``.loc``).

So give it a whirl. If you find missing / unclear tests, let's fix em. The more comprehensive the better.

jreback commented:
cc @shoyer @jorisvandenbossche @TomAugspurger @wesm

jorisvandenbossche commented:
Great!

Quick note: I think we should also raise on ``[]`` when this does positional indexing (eg in slicing) and using floats (just as `iloc`). Eg `s[1:2.5]` raised in 0.17.1, but now works with this PR.

I will try to take a more detailed look later this week.

jreback commented:
```
In [1]: s = Series([1,2,3])

In [2]: s[1:2.5]
TypeError: cannot do slice stop value indexing on <class 'pandas.core.index.Int64Index'> with these indexers [2.5] of <type 'float'>

In [3]: s.loc[1:2.5]
Out[3]: 
1    2
2    3
dtype: int64
```

why are you calling out this a special case? This is the same path as ``.loc``.

jorisvandenbossche commented:
I get 

```
In [48]: s = Series([1,2,3])

In [49]: s[1:2.5]
Out[49]: array([2], dtype=int64)
```

using your branch (so something else as you show in your last comment)

jorisvandenbossche commented:
Which now I note gives an array instead of series, which is also odd

jreback commented:
@jorisvandenbossche yeh something odd going on. will update soon. In any event I don't see a reason to make this any different it is very consistent with the other indexers.

jorisvandenbossche commented:
Very strange, it seems to depend on the values in the index:

```
In [14]: s = Series([1,2,3])

In [15]: s[1:2.5]
Out[15]: array([2], dtype=int64)

In [16]: s = Series([1,2,3], index=[1,2,3])

In [17]: s[1:2.5]
Out[17]:
2    2
dtype: int64
```

jreback commented:
@jorisvandenbossche ok, adding tests for these off integers.

jorisvandenbossche commented:
But to come back to what I originally wanted to say: raising in []. 
You gave above the example (https://github.com/pydata/pandas/pull/12370#issuecomment-187725450):

```
In [1]: s = Series([1,2,3])

In [2]: s[1:2.5]
TypeError: cannot do slice stop value indexing on <class 'pandas.core.index.Int64Index'> with these indexers [2.5] of <type 'float'>

In [3]: s.loc[1:2.5]
Out[3]: 
1    2
2    3
dtype: int64
```

This *raises* with floats when slicing in `[]`. And this is what I meant that I think should happen, but it is not what currently happens in the PR.

jreback commented:
oh, I just generalized things. This is particular was not exactly tested :< oh
ok in agreement that this is NOT a special case then? gr8

jorisvandenbossche commented:
@jreback yep, ok, in agreement it is not a special case and should stay as it was in 0.17.1 (your example). The confusion was that is actually behaved differently in the PR, which I tried to point out :-)

jreback commented:
@jorisvandenbossche no, my example is from 0.17.1. But I am NOT treating it as a special case, which means this WILL work the same as ``.loc`` (aside fro the ``np.array(..)`` issue which I have to track).

why should this raise?

jorisvandenbossche commented:
> why should this raise?

Because `[]` is not like `loc` in this case, but like `iloc` (which raises)

jreback commented:
@jorisvandenbossche no i disagree, this is label based. Again this is introducing special cases where none should exist.

jorisvandenbossche commented:
I am afraid that is not label based, but positional, and this is a special case that is already present ...

```
In [12]: pd.__version__
Out[12]: u'0.17.1'

In [13]: s = pd.Series(range(2,6), index=range(2,6))

In [14]: s
Out[14]:
2    2
3    3
4    4
5    5
dtype: int64

In [15]: s[2:4]
Out[15]:
4    4
5    5
dtype: int64

In [16]: s.loc[2:4]
Out[16]:
2    2
3    3
4    4
dtype: int64
```

jreback commented:
this is exactly the ``.ix`` path

jorisvandenbossche commented:
I am not saying this is ideal :-) But it is in any case positional and not label based, and that is not something we can change lightly

> this is exactly the .ix path

that does not seems to be the case, as `ix` *is* indeed label based in this case

```
In [17]: s.ix[2:4]
Out[17]:
2    2
3    3
4    4
dtype: int64
```

jreback commented:
my point is that this is EXACTLY THE SAME PATH. It IS not a special case in this branch, nor should it. 

IIRC that was accidental before.

jreback commented:
ok, you are right it IS positional. So this seems entirely reasonable then. Yes would be a slight API change, becuase now we accept something we don't before, but it IS consistent. Yes it IS like ``.iloc``, but so is ``[]`` in this case. But again removing this is a special case.

```
In [1]: s = pd.Series([1,2,3],index=[4,5,6])

In [2]: s[1:3]
Out[2]: 
5    2
6    3
dtype: int64

In [3]: s[1.5:3]
Out[3]: 
5    2
6    3
dtype: int64

In [4]: s.ix[1.5:3]
Out[4]: Series([], dtype: int64)

In [5]: s.loc[1.5:3]
Out[5]: Series([], dtype: int64)
```

jorisvandenbossche commented:
> ok, you are right it IS positional. So this seems entirely reasonable then. 

I am not sure how to interpret this (is it my or your point that is reasonable :-) ?) Can you show what you think the behaviour should be? (or is it like your last code example?)

My point of vue: independent of it is in `[]/ix/iloc`, we have the concept of 'positional' indexing. For `[]/ix` it depends on some factors if it will be positional or not. But once we have positional indexing, I would like to see the behaviour consistent for all indexer types within the concept of positional indexing. I think it would be really unfortunate that we introduce inconsistencies within that concept.

jreback commented:
@jorisvandenbossche ok finally had time to update this.

lmk what you think

jorisvandenbossche commented:
@jreback I still think positional indexing should always raise when using floats, also in slicing in [] (as it already raised in 0.17). See my last post above. 
What is your rationale to make an exception and now allow it?

Will try to take a more detailed look tomorrow

jreback commented:
can u five an explicit example where u think I am doing positional indexing with floats? for scalars and slicing

jorisvandenbossche commented:
```
In [13]: pd.__version__
Out[13]: '0.18.0rc1+79.gff90e0a'

In [14]: s = pd.Series(range(2,6), index=range(2,6))

In [17]: s.loc[2:4]
Out[17]:
2    2
3    3
4    4
dtype: int64

In [18]: s[2:4]
Out[18]:
4    4
5    5
dtype: int64

In [19]: s[2:4.0]   <---------- this is positional indexing with floats -> so should raise?
Out[19]:
4    4
5    5
dtype: int64
```
vs

```
In [6]: pd.__version__
Out[6]: u'0.17.1'

In [7]:  s = pd.Series(range(2,6), index=range(2,6))

In [8]: s[2:4.0]          <------------ does raise in 0.17.1
TypeError: cannot do slice stop value indexing on <class 'pandas.core.index.Int6
4Index'> with these indexers [4.0] of <type 'float'>
```

jreback commented:
ok let me adjust that

jreback commented:
@jorisvandenbossche ok updated. I *think* that we are now consistent with 0.17.1 (and now fully tested).

jorisvandenbossche commented:
This looks good! Thanks a lot

jorisvandenbossche commented:
An edge case: if you have mixed dtype index, you get some inconsistencies:

```
In [48]: s2 = pd.Series([1,2,3], index=['a', 'b', 'c'])

In [49]: s3 = pd.Series([1,2,3], index=['a', 'b', 1.5])

In [50]: s2[1.0]
TypeError: cannot do label indexing on <class 'pandas.indexes.base.Index'> with
these indexers [1.0] of <type 'float'>

In [51]: s3[1.0]
Out[51]: 2
```

So the second case (`s3[1.0]`) appears to do positional indexing with a float (which should raise?)

jorisvandenbossche commented:
Actually, the fact that `s2[1.0]` (and also `s2.ix[1.0]`) in the above example raises, is an API change (previously it was coerced to integer and positional indexing was done), so we should mention this in the whatsnew docs I think

jreback commented:
hmm, I'll put in a fix for that. was not the intention to change the API (at all)

jorisvandenbossche commented:
No, I think it is a good change (although maybe we should have raised a FutureWarning for that ..). With a full string index, `s2[1.0]` is either a KeyError, or either is tries to do positional but then should raise because it is a float (and positional is strictly with integers)

jreback commented:
neh, this is really odd ``.ix`` fails here as well

jorisvandenbossche commented:
And they raised a warning, so I think it is certainly OK that they now both raise

jreback commented:
ok, will push a fix in a minute

jorisvandenbossche commented:
So the only issue here , then, is for the mixed dtype index, where it should also raise.

jreback commented:
actually, I could make a case for ``s3.ix[1.0]`` just being a ``KeyError`` as its label based (as the index is mixed)

jreback commented:
I added tests, but I didnt' actually change any code. lmk what you think.

jorisvandenbossche commented:
> actually, I could make a case for s3.ix[1.0] just being a KeyError as its label based (as the index is mixed)

+1

jreback commented:
@jorisvandenbossche 

on something like this. (from 0.17.1)

```
In [3]: In [48]: s2 = pd.Series([1,2,3], index=['a', 'b', 'c'])

In [4]: s2.loc[1.0]
KeyError: 'the label [1] is not in the [index]'
```

but we are a bit more exact in that this is REALLY a ``TypeError``, or is this just confusing? e.g. we KNOW by the dtype that some values are simply an error (floats in a DatetimeIndex for example). For a mixed index, we are not sure and HAVE to raise a ``KeyError``. But what about a pure-string index (which we currently raise a ``KeyError``), that should actually be a ``TypeError``?

jorisvandenbossche commented:
If we had something like a StringIndex, it would be clear that a float is not possible. But for a plain Index, although it has only string values, .. Hmm, it's a bit a dubious case of course. If it was a KeyError previously, I would maybe keep it that way? (unless that makes it more complicated)

jreback commented:
yeah just going to keep it. pushing in a moment.

jreback commented:
ok all fixed up. simplified the logic a bit so its pretty clear.

jreback commented on pandas/indexes/base.py:
```diff
@@ -966,44 +968,41 @@ def _convert_scalar_indexer(self, key, kind=None):
-        if self.is_integer() or is_index_slice:
-            return key
-        return self._convert_slice_indexer(key)
+            if kind in ['getitem', 'ix'] and is_float(key):
```
This is where it aises a ``TypeError`` here if its label based but not the right type.
